<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <!-- カメラの寸法 (Camera Dimensions in meters) -->
    <xacro:property name="camera_length" value="0.03"/>
    <xacro:property name="camera_width" value="0.10"/>
    <xacro:property name="camera_height" value="0.03"/>
    <xacro:property name="camera_mass" value="0.05"/>
    
    <!-- カメラの位置 (Camera Position) -->
    <xacro:property name="camera_offset_x" value="${chassis_radius + camera_length/2}"/>
    <xacro:property name="camera_offset_z" value="${chassis_height/2 + camera_height/2 + 0.02}"/>
    
    <!-- カメラパラメータ (Camera Parameters) -->
    <xacro:property name="camera_width_px" value="640"/>
    <xacro:property name="camera_height_px" value="480"/>
    <xacro:property name="camera_fov" value="1.3962634"/>  <!-- 80 degrees -->
    <xacro:property name="camera_fps" value="30"/>

    <!-- カメラリンク (Camera Link) -->
    <link name="camera_link">
        <visual>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="${camera_length} ${camera_width} ${camera_height}"/>
            </geometry>
            <material name="red"/>
        </visual>
        
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
                <box size="${camera_length} ${camera_width} ${camera_height}"/>
            </geometry>
        </collision>
        
        <xacro:inertial_box mass="${camera_mass}" x="${camera_length}" y="${camera_width}" z="${camera_height}">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>

    <!-- カメラジョイント (Camera Joint) -->
    <joint name="camera_joint" type="fixed">
        <parent link="chassis"/>
        <child link="camera_link"/>
        <origin xyz="${camera_offset_x} 0 ${camera_offset_z}" rpy="0 ${10*pi/180} 0"/>
    </joint>

    <!-- カメラ光学フレーム (Camera Optical Frame) -->
    <link name="camera_link_optical"/>

    <joint name="camera_optical_joint" type="fixed">
        <parent link="camera_link"/>
        <child link="camera_link_optical"/>
        <origin xyz="0 0 0" rpy="${-pi/2} 0 ${-pi/2}"/>
    </joint>

    <!-- Gazeboカメラプラグイン (Gazebo Camera Plugin) -->
    <gazebo reference="camera_link">
        <material>Gazebo/Red</material>
        
        <sensor name="camera_sensor" type="camera">
            <update_rate>${camera_fps}</update_rate>
            <visualize>true</visualize>
            
            <camera name="camera">
                <horizontal_fov>${camera_fov}</horizontal_fov>
                <image>
                    <width>${camera_width_px}</width>
                    <height>${camera_height_px}</height>
                    <format>R8G8B8</format>
                </image>
                <clip>
                    <near>0.1</near>
                    <far>100</far>
                </clip>
                <noise>
                    <type>gaussian</type>
                    <mean>0.0</mean>
                    <stddev>0.007</stddev>
                </noise>
            </camera>
            
            <plugin name="camera_controller" filename="libgazebo_ros_camera.so">
                <ros>
                    <namespace>/</namespace>
                    <remapping>~/image_raw:=/camera/image_raw</remapping>
                    <remapping>~/camera_info:=/camera/camera_info</remapping>
                </ros>
                <camera_name>camera</camera_name>
                <frame_name>camera_link_optical</frame_name>
                <hack_baseline>0.07</hack_baseline>
            </plugin>
        </sensor>
    </gazebo>

</robot>